[RFC] Add http_proxy parameter to allow specifying per-job HTTP proxy servers

wrouesnel opened the pull request:
RFC for adding the option to use HTTP proxies for scraping endpoints for specific scrape-configs.

Solves the problem of a Prometheus server that might be mostly capable of scraping it's hosts on an internal network but which needs to reach one or several external servers which don't warrant (or which you can't) deploy a separate instance to federate with.

fabxc commented:
Thanks and sorry for the long silence.

We would like to support proxying. Via #921 your changes are unfortunately outdated. If you could rebase and adjust as necessary, I will review as soon as possible.

wrouesnel commented:
I've updated the code to take account of the new changes.

In view that it might want to do other things, I've renamed the parameter proxy_url in-case other proxy functions (maybe SOCKS) should be added in the future.

Internally it's a little different - rather then having a separate constructor, I've just added a proxy url parameter to all the relevant client.go calls, which seems like a better way to structure it.

fabxc commented on config/config.go:
```diff
@@ -69,6 +69,7 @@ var (
 		MetricsPath: "/metrics",
 		Scheme:      "http",
 		HonorLabels: false,
+		Proxy: URL{},
```
No need to default zero values unless it's worth to explicitly state it like in `HonorLabels`.
But not seeing a proxy in this struct intuitively means that there is none.

fabxc commented on config/config.go:
```diff
@@ -100,6 +101,35 @@ var (
+	if u.URL != nil {
+		return u.String(), nil
+	} else {
+		return nil,nil
```
Run `go fmt` please.

fabxc commented on config/config.go:
```diff
@@ -237,6 +267,8 @@ type ScrapeConfig struct {
 	// The client cert authentication credentials for the targets.
 	ClientCert *ClientCert `yaml:"client_cert,omitempty"`
+        // HTTP proxy server to use to connect to the targets.
+	Proxy URL `yaml:"proxy_url,omitempty"`
```
Let's keep the member name and YAML name in sync and use `ProxyURL`.

fabxc commented on util/httputil/client.go:
```diff
@@ -26,14 +27,16 @@ func NewClient(rt http.RoundTripper) *http.Client {
 // NewDeadlineRoundTripper returns a new http.RoundTripper which will time out
 // long running requests.
-func NewDeadlineRoundTripper(timeout time.Duration) http.RoundTripper {
+func NewDeadlineRoundTripper(timeout time.Duration, proxyUrl *url.URL) http.RoundTripper {
```
`proxyURL`

fabxc commented:
Looks good otherwise.

wrouesnel commented:
A little gorename magic and it should be ready.

fabxc commented:
Thanks a lot!

TheTincho commented:
I would love to have support for SOCKS proxies too, as I find myself having to use SSH tunnels to connect to remote targets,  so to use this I will need to setup an http proxy in front of the ssh tunnel
